File: packages/gateway/src/observability.ts
import { bearerFromHeader } from './auth.js';
import type { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';
import {
  Registry,
  Counter,
  Histogram,
  Gauge,
  collectDefaultMetrics,
} from 'prom-client';

const SERVICE = 'lotaru';

const register = new Registry();
collectDefaultMetrics({ register, prefix: 'lotaru_' });

const httpRequests = new Counter({
  name: 'http_requests_total',
  help: 'Total HTTP requests',
  labelNames: ['service', 'method', 'route', 'status_class'] as const,
  registers: [register],
});

const httpDuration = new Histogram({
  name: 'http_request_duration_seconds',
  help: 'HTTP request duration',
  labelNames: ['service', 'method', 'route', 'status_class'] as const,
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10],
  registers: [register],
});

const httpInFlight = new Gauge({
  name: 'http_requests_in_flight',
  help: 'In-flight HTTP requests',
  labelNames: ['service'] as const,
  registers: [register],
});

function statusClass(code: number): string {
  if (code >= 500) return '5xx';
  if (code >= 400) return '4xx';
  if (code >= 300) return '3xx';
  if (code >= 200) return '2xx';
  return '1xx';
}

function normalizeRoute(url: string): string {
  const path = url.split('?')[0] ?? '/';
  return path
    .split('/')
    .map((p) => {
      if (p.length === 0) return p;
      if (/^[0-9a-f-]{8,}$/i.test(p) || /^\d+$/.test(p)) return ':id';
      return p;
    })
    .join('/');
}

function clientIp(req: FastifyRequest): string {
  const xf = req.headers['x-forwarded-for'];
  if (typeof xf === 'string' && xf.length > 0) {
    return xf.split(',')[0]?.trim() ?? req.ip;
  }
  if (Array.isArray(xf) && xf[0]) {
    return xf[0].split(',')[0]?.trim() ?? req.ip;
  }
  return req.ip;
}

export async function registerObservability(app: FastifyInstance): Promise<void> {
  app.get('/metrics', async (req, reply) => {
    const expected = process.env['METRICS_TOKEN'];
    if (expected === undefined || expected.length === 0) {
      return reply.code(404).send();
    }
    const got = bearerFromHeader(req.headers.authorization);
    if (got !== expected) {
      return reply.code(401).send({ error: 'unauthorized' });
    }
    reply.header('Content-Type', register.contentType);
    return register.metrics();
  });

  app.addHook('onRequest', async (req) => {
    if (req.url.startsWith('/metrics') || req.url.startsWith('/healthz')) {
      return;
    }
    httpInFlight.inc({ service: SERVICE });
    (req as FastifyRequest & { __obsStart?: bigint }).__obsStart = process.hrtime.bigint();
  });

  app.addHook('onResponse', async (req: FastifyRequest, reply: FastifyReply) => {
    if (req.url.startsWith('/metrics') || req.url.startsWith('/healthz')) {
      return;
    }
    httpInFlight.dec({ service: SERVICE });
    const start = (req as FastifyRequest & { __obsStart?: bigint }).__obsStart;
    const durMs =
      start === undefined ? 0 : Number(process.hrtime.bigint() - start) / 1e6;
    const route = normalizeRoute(req.url);
    const sc = statusClass(reply.statusCode);
    httpRequests.inc({ service: SERVICE, method: req.method, route, status_class: sc });
    httpDuration.observe(
      { service: SERVICE, method: req.method, route, status_class: sc },
      durMs / 1000,
    );
    const line = {
      msg: 'http_access',
      service: SERVICE,
      client_ip: clientIp(req),
      method: req.method,
      path: req.url.split('?')[0] ?? '/',
      status: reply.statusCode,
      duration_ms: Math.round(durMs * 100) / 100,
      request_id: req.id,
    };
    process.stdout.write(`${JSON.stringify(line)}\n`);
  });
}

File: auth/internal/httpapi/observability.go
package httpapi

import (
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/go-chi/chi/v5/middleware"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var (
	httpRequests = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "http_requests_total",
		Help: "Total HTTP requests",
	}, []string{"service", "method", "route", "status_class"})

	httpDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "http_request_duration_seconds",
		Help:    "HTTP request duration",
		Buckets: prometheus.DefBuckets,
	}, []string{"service", "method", "route", "status_class"})

	httpInFlight = promauto.NewGauge(prometheus.GaugeOpts{
		Name: "http_requests_in_flight",
		Help: "In-flight HTTP requests",
		ConstLabels: prometheus.Labels{
			"service": "auth",
		},
	})
)

func statusClass(code int) string {
	switch {
	case code >= 500:
		return "5xx"
	case code >= 400:
		return "4xx"
	case code >= 300:
		return "3xx"
	case code >= 200:
		return "2xx"
	default:
		return "1xx"
	}
}

func normalizeRoute(r *http.Request) string {
	if rctx := chi.RouteContext(r.Context()); rctx != nil {
		if p := rctx.RoutePattern(); p != "" {
			return p
		}
	}
	path := r.URL.Path
	if path == "" {
		return "/"
	}
	parts := strings.Split(path, "/")
	for i, p := range parts {
		if p == "" {
			continue
		}
		if len(p) > 20 || looksLikeID(p) {
			parts[i] = ":id"
		}
	}
	return strings.Join(parts, "/")
}

var sensitiveQueryKeys = map[string]struct{}{
	"token":          {},
	"access_token":   {},
	"refresh_token":  {},
	"code":           {},
	"key":            {},
	"secret":         {},
	"password":       {},
	"authorization":  {},
	"api_key":        {},
	"client_secret":  {},
}

func sanitizeQuery(raw string) string {
	if raw == "" {
		return ""
	}
	q := raw
	if idx := strings.Index(raw, "?"); idx >= 0 {
		q = raw[idx+1:]
	}
	if q == "" {
		return ""
	}
	vals, err := url.ParseQuery(q)
	if err != nil {
		if len(q) > 500 {
			return q[:500] + "…"
		}
		return q
	}
	out := url.Values{}
	for key, items := range vals {
		redact := false
		if _, ok := sensitiveQueryKeys[strings.ToLower(key)]; ok {
			redact = true
		}
		for _, item := range items {
			if redact {
				out.Add(key, "[redacted]")
				continue
			}
			if len(item) > 120 {
				out.Add(key, item[:40]+"…")
				continue
			}
			out.Add(key, item)
		}
	}
	text := out.Encode()
	if len(text) > 500 {
		return text[:500] + "…"
	}
	return text
}

func looksLikeID(s string) bool {
	if len(s) >= 32 {
		return true
	}
	_, err := strconv.ParseUint(s, 10, 64)
	return err == nil
}

func clientIP(r *http.Request) string {
	candidates := []string{
		r.Header.Get("CF-Connecting-IP"),
		r.Header.Get("True-Client-IP"),
		r.Header.Get("X-Real-IP"),
	}
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		for _, part := range strings.Split(xff, ",") {
			candidates = append(candidates, strings.TrimSpace(part))
		}
	}
	for _, raw := range candidates {
		if raw == "" {
			continue
		}
		host := raw
		if h, _, err := net.SplitHostPort(raw); err == nil {
			host = h
		}
		ip := net.ParseIP(host)
		if ip == nil || ip.IsLoopback() || ip.IsPrivate() || ip.IsLinkLocalUnicast() || ip.IsUnspecified() {
			continue
		}
		return host
	}
	host, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		return r.RemoteAddr
	}
	return host
}

type accessLine struct {
	Msg        string  `json:"msg"`
	Service    string  `json:"service"`
	ClientIP   string  `json:"client_ip"`
	Method     string  `json:"method"`
	Path       string  `json:"path"`
	Route      string  `json:"route"`
	Query      string  `json:"query"`
	Status     int     `json:"status"`
	DurationMs float64 `json:"duration_ms"`
	RequestID  string  `json:"request_id"`
}

func accessAndMetrics(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/metrics" || r.URL.Path == "/healthz" {
			next.ServeHTTP(w, r)
			return
		}
		httpInFlight.Inc()
		start := time.Now()
		ww := middleware.NewWrapResponseWriter(w, r.ProtoMajor)
		defer func() {
			httpInFlight.Dec()
			dur := time.Since(start)
			code := ww.Status()
			if code == 0 {
				code = http.StatusOK
			}
			route := normalizeRoute(r)
			sc := statusClass(code)
			httpRequests.WithLabelValues("auth", r.Method, route, sc).Inc()
			httpDuration.WithLabelValues("auth", r.Method, route, sc).Observe(dur.Seconds())
			line := accessLine{
				Msg:        "http_access",
				Service:    "auth",
				ClientIP:   clientIP(r),
				Method:     r.Method,
				Path:       r.URL.Path,
				Route:      route,
				Query:      sanitizeQuery(r.URL.RawQuery),
				Status:     code,
				DurationMs: float64(dur.Milliseconds()),
				RequestID:  middleware.GetReqID(r.Context()),
			}
			b, err := json.Marshal(line)
			if err == nil {
				_, _ = fmt.Fprintln(os.Stdout, string(b))
			}
		}()
		next.ServeHTTP(ww, r)
	})
}

func metricsHandler() http.Handler {
	return promhttp.Handler()
}

func (s *Server) metricsAuth(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		expected := strings.TrimSpace(s.cfg.MetricsToken)
		if expected == "" {
			writeError(w, http.StatusNotFound, "not_found", "metrics unavailable")
			return
		}
		authz := strings.TrimSpace(r.Header.Get("Authorization"))
		if authz != "Bearer "+expected {
			writeError(w, http.StatusUnauthorized, "unauthorized", "metrics requires authorization")
			return
		}
		next.ServeHTTP(w, r)
	})
}
